src/repo.rs

```diff
@@ -0,0 +1,59 @@
+use crate::errors::*;
+
+use miette::miette;
+
+#[derive(Debug)]
+pub enum GithubRepoInput {
+    Url(String),
+    Ssh(String),
+}
+
+impl GithubRepoInput {
+    pub fn new(repo_string: String) -> Result<Self> {
+        // Handle git+https just the same as https
+        if repo_string.starts_with("https") || repo_string.starts_with("git+https") {
+            Ok(Self::Url(repo_string))
+        } else if repo_string.starts_with("git@") {
+            Ok(Self::Ssh(repo_string))
+        } else {
+            let err = AxoprojectError::UnknownRepoStyle { url: repo_string };
+            Err(err)
+        }
+    }
+
+    pub fn normalize(self) -> Result<String> {
+        match self {
+            Self::Url(s) => Ok(Self::normalize_url(s)?),
+            Self::Ssh(s) => Ok(Self::normalize_ssh(s)?),
+        }
+    }
+
+    fn normalize_url(repo_string: String) -> Result<String> {
+        Ok(Self::remove_git_suffix(repo_string.to_string()))
+    }
+
+    fn normalize_ssh(repo_string: String) -> Result<String> {
+        let core = Self::remove_git_suffix(Self::replace_git_prefix(repo_string.clone())?);
+        Ok(Self::remove_git_suffix(core))
+    }
+
+    fn replace_git_prefix(s: String) -> Result<String> {
+        let prefix = "git@github.com:";
+        if s.starts_with(prefix) {
+            Ok(s.replace(prefix, "https://github.com/"))
```

Comment:
Enterprise GitHub would follow the same format, but I don't think we can quite guarantee what other platforms use. I can put in an assumption everything follows the GitHub format, and it would at least support Bitbucket, etc, but it'll for sure result in weird URLs in some other cases.

Reply:
we only support github right now and don't have plans to expand any time soon, so i think we should make this work *just* for github and hard error on other things.

Reply:
hard error for github enterprise as well? (could be an arbitrary format domain)

https://github.com/axodotdev/cargo-dist/issues/185